target/linux/ath79/dts/ar9132_buffalo_wzr-hp-g300nh.dts

```diff
@@ -0,0 +1,260 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "ar9132.dtsi"
+
+/ {
+	compatible = "buffalo,wzr-hp-g300nh", "qca,ar9132";
+	model = "Buffalo WZR-HP-G300NH";
+
+	aliases {
+		led-boot = &security;
+		led-failsafe = &diag;
+		led-upgrade = &diag;
+	};
+
+	clock40mhz: ref {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <40000000>;
+	};
+
+	gpio2: gpio-extender {
+		compatible = "nxp,74hc153-gpio";
+		gpio-controller;
+		ngpios = <8>;
+		#gpio-cells = <2>;
+
+		// GPIOs used by this node
+		gpio-s0 = <&gpio 9 GPIO_ACTIVE_HIGH>;
+		gpio-s1 = <&gpio 11 GPIO_ACTIVE_HIGH>;
+		gpio-1y = <&gpio 12 GPIO_ACTIVE_HIGH>;
+		gpio-2y = <&gpio 14 GPIO_ACTIVE_HIGH>;
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+
+		poll-interval = <20>;
```

Comment:
Please use gpio-keys instead and remove poll-interval.

Reply:
Can you please elaborate on this? Why not gpio-keys-polled? Neither of these seem to work in master, though. With gpio-keys-polled I end up in failsafe always and gpio-keys cannot register gpios for its use. Something has apparently changed between master and 19.07 but I cannot put my finger on it.

Reply:
ath79 and most ramips devices have been switches to gpio-keys a while ago:
https://github.com/openwrt/openwrt/commit/8961d2268e3c196ed0bcdb7eded622c434982925
https://github.com/openwrt/openwrt/commit/556ff0987539271647535094309605433f2aa96f

The interrupt-based driver is expected to be slightly better, though there are many devices where both variants don't work properly. Nevertheless, we haven't found a device where the "new" driver makes it worse.

Therefore, gpio-keys is preferred, unless you have ath9k-leds which need the polling driver anyway or you are the first person where the polled driver work reproducibly better.

If you boot into failsafe, this is typically an indication for a wrong GPIO number or one with active_high/active_low set wrong.

Reply:
I would be happy to use gpio-keys but I get this in the log:

> [ 3.498802] gpio-keys keys: failed to get irq for gpio:504
> [ 3.504409] gpio-keys keys: failed to get irq for gpio:505
> [ 3.509944] gpio-keys keys: failed to get irq for gpio:506
> [ 3.515479] gpio-keys keys: failed to get irq for gpio:507
> [ 3.521017] gpio-keys keys: failed to get irq for gpio:509
> [ 3.526554] gpio-keys keys: failed to get irq for gpio:510
> [ 3.532091] gpio-keys keys: failed to get irq for gpio:511

Also, I found reason why I end up in failsafe every time with gpio-keys-polled. By reverting this change things start to work again. https://github.com/openwrt/openwrt/commit/174ff7d754b272cbf6169bfbb245dbd12b392830